File: Two_Pointers/016.3Sum-Closet/016.3Sum-Closet.ts
function threeSumClosest(nums: number[], target: number): number {
  let ret = 0;
  let gapMin = Number.MAX_SAFE_INTEGER;
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; i++) {
    let left = i + 1;
    let right = nums.length - 1;
    while (left < right) {
      const sumOfThree = nums[i] + nums[left] + nums[right];
      const gap = Math.abs(target - sumOfThree);
      if (gap < gapMin) {
        gapMin = gap;
        ret = sumOfThree;
      }

      if (sumOfThree > target) {
        right--;
      } else {
        left++;
      }
    }
  }
  return ret;
}

File: Two_Pointers/016.3Sum-Closet/016.3Sum-Closet.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int minGap = Integer.MAX_VALUE;
        
        int result = -1;
        for (int i = 0; i < nums.length; i++) {
            int left = i + 1;
            int right = nums.length - 1;
            
            while (left < right) {
                int sum = nums[left] + nums[right] + nums[i];

                if (sum == target) {
                    return sum;
                } else {
                    int gap = Math.abs(target - sum);
                    if (gap < minGap) {
                        minGap = gap;
                        result = sum;
                    }
                    
                    if (sum > target) {
                        right -= 1;
                    } else {
                        left += 1;
                    }
                }
            }
        }
        return result;

    }
}
